unit/solvers/smt2_incremental/smt_terms.cpp

```diff
@@ -0,0 +1,221 @@
+// Author: Diffblue Ltd.
+
+#include <testing-utils/use_catch.h>
+
+#include <solvers/smt2_incremental/smt_terms.h>
+
+#include <util/mp_arith.h>
+
+#include <type_traits>
+
+TEST_CASE("smt_bool_literal_termt upcasts.", "[core][smt2_incremental]")
+{
+  smt_termt term{smt_bool_literal_termt{true}};
+  term = smt_bool_literal_termt{false};
+}
+
+TEST_CASE("Test smt_termt.pretty is accessible.", "[core][smt2_incremental]")
+{
+  const smt_termt bool_term{smt_bool_literal_termt{true}};
+  REQUIRE(
+    bool_term.pretty(0, 0) ==
+    "smt_bool_literal_term\n"
+    "  * type: smt_bool_sort\n"
+    "  * value: 1");
+}
+
+TEST_CASE(
+  "Test smt_bool_literal_termt has bool sort.",
+  "[core][smt2_incremental]")
+{
+  REQUIRE(smt_bool_literal_termt{true}.get_sort() == smt_bool_sortt{});
+}
+
+TEST_CASE(
+  "Test smt_bool_literal_termt value getter.",
+  "[core][smt2_incremental]")
+{
+  REQUIRE(smt_bool_literal_termt{true}.value());
+  REQUIRE_FALSE(smt_bool_literal_termt{false}.value());
+}
+
+TEST_CASE("smt_not_termt upcasts.", "[core][smt2_incremental]")
+{
+  smt_termt not_term{smt_not_termt{smt_bool_literal_termt{false}}};
+  not_term = smt_not_termt{smt_bool_literal_termt{true}};
+}
+
+TEST_CASE("smt_not_termt sort.", "[core][smt2_incremental]")
+{
+  REQUIRE(
+    smt_not_termt{smt_bool_literal_termt{true}}.get_sort() == smt_bool_sortt{});
+}
+
+TEST_CASE("smt_not_termt operand getter.", "[core][smt2_incremental]")
+{
+  const smt_bool_literal_termt bool_term{true};
+  const smt_not_termt not_term{bool_term};
+  REQUIRE(not_term.operand() == bool_term);
+}
+
+TEST_CASE("smt_identifier_termt construction", "[core][smt2_incremental]")
+{
+  cbmc_invariants_should_throwt invariants_throw;
+  CHECK_NOTHROW(smt_identifier_termt{"foo bar", smt_bool_sortt{}});
+  CHECK_THROWS(smt_identifier_termt{"|foo bar|", smt_bool_sortt{}});
+  CHECK_THROWS(smt_identifier_termt{"foo\\ bar", smt_bool_sortt{}});
+}
+
+TEST_CASE("smt_identifier_termt upcasts.", "[core][smt2_incremental]")
```

Comment:
No actual `CHECK` inside this test case, I guess it's a style choice?

Reply:
Yeah, I wrote these upcast tests when I was initially working on the accessibility, at that time I just wanted to be sure that the upcasts compiled. However now that the rest of the tests are in place and full of upcasts, I am not sure that this has value. Therefore I might just remove all the upcast tests.

Reply:
I have now removed these tests, as I was unconvinced of the value they added.